### ISSUE TITLE
Projects people can help out with

### ISSUE THREAD

**Issue by nathansobo** [nathansobo/text-transform#1]
I need to do a more detailed design write-up on what I'm thinking for this library, but here's a list of projects an adventurous contributor (/cc @as-cii) could help out with.
- **Add indent support on the soft-wraps layer.** @as-cii just did this on master so it should be fresh. It probably requires `getNextRegion` in our transform objects to be replaced with `getRegions` so they can return multiple.
- **Implement a new layer**
  - **Soft tabs**
    This would group leading whitespace into "clipped" (formerly know as "atomic") tokens.
  - **Folds**
    This requires markers to be implemented first, but it's a pretty interesting problem.
  - **Syntactic scope annotation**
    This would replace the strings we use as our content type with attributed strings, a data type that mixed scope start/end tags with string content and abstracted over the presence of the scope tags in various operations unless we explicitly asked for them. This also requires the state of the parser to be conveyed between lines and considered when invalidating. It should also be asynchronous. Ultimately this layer will be a place for a lot of optimization work, such as potentially moving the parser to the native layer, but just getting it into the layered scheme would be a good start.
  - **Invisibles**
    First you'd need to reintroduce control characters like `\n` and `\t` that I mistakenly decided to strip out on the lines and hard tabs layer. Then a higher layer could replace them with whitespace or their invisible equivalents. We should probably allow multi-character invisibles since some people have asked for them for some strange reason.
- **Add randomized spec coverage** Similar to our approach with `TextEditorPresenter`, this library could probably benefit from randomized fuzz testing. Create a buffer, randomly mutate it, then compare its layers with the layers of a newly constructed buffer with identical content. This can help us catch bugs in our incremental update logic.
- **Implement markers (linearly)**
  This would associate marker end-points directly with regions. If a marker starts in a region, we would track the traversal for the marker's relative start point inside the region. If the marker spans across a region, we'd mark whether the marker spans over the left endpoint, the right endpoint, or both. A first step would then implement stabbing queries against our region storage in a way that could be directly adapted to a transition to storing regions in a b-tree instead of an array.
- **Implement marker observation windows**
  Right now we subscribe to every marker directly. We should instead subscribe to an observation window that reports when markers are added, removed, or moved relative to that window.
- **Switch storage to a counted b-tree**
  The timing of this is delicate because it adds complexity to the implementation of everything else. However, if you don't get too attached to any specific code we can start learning by implementing the features we have so far on a branch. There may be churn though as the basic structure evolves. Subprojects in the b-tree project:
  - Marker storage on leaf nodes.
  - Index maintenance for traversals in source and target coordinate spaces, absolute offset in terms of regions, and maximal the row/column position for the file.
  - What arity of tree should we use? My gut tells me a 2-3-4 tree is fine because we're in memory, but I'm not deeply experienced in this area.

There's probably a lot more, but that should get you started. This project is extremely young, so you'll need to anticipate keeping pace with a lot of change. It's possible that merging pull request will need to be deferred based on other work that's taking place, which means potentially a lot of branch maintenance. However, with that understanding, there's a lot of learning and implementation work that can probably happen in parallel.
